IPIP/0000-double-hash-dht.md

```diff
@@ -0,0 +1,339 @@
+- **Provider Store** is the data structure on the DHT Servers used to store the Provider Records. Its structure is a nested dictionary/map: `HASH2` -> `ServerKey` -> [`CPPeerID`, `EncPeerID`, `TS`, `Signature`]. There is only one single correct `ServerKey` for each `HASH2`. However, any peer can forge a valid Publish request (with invalid `EncPeerID` but valid `Signature`) undetected by the DHT Server. The DHT server isn't able to distinguish which `ServerKey` is correct as it doesn't have the knowledge of `MH`, hence it has to keep both and serve both upon request for `HASH2`.
+
+**Magic Values**
+- bytes("CR_DOUBLEHASH")
+- bytes("CR_SERVERKEY")
+- AESGCM varint: `TODO` <!-- TODO: add varint draft -->
+- Double SHA256 varint: `DBL_SHA2_256 = 86`
+- Max number of Provider Records returned by a DHT Server for a single request: `128` <!-- TODO: define number-->
+
+### Current DHT
+
+The following process describes the event of a client looking up a CID in the IPFS DHT:
+1. Client computes `Hash(MH)` (`MH` is the MultiHash included in the CID).
+2. Client looks for the closest peers to `Hash(MH)` in XOR distance in its Routing Table.
+3. Client sends a DHT lookup request for `CID` to these DHT servers.
+4. Upon receiving the request, the DHT servers search if there is an entry for `MH` in their Provider Store. If yes, go to 10. Else continue.
+5. DHT servers compute `Hash(MH)`.
+6. DHT servers find the 20 closest peers to `Hash(HM)` in XOR distance in their Routing Table.
+7. DHT servers return the 20 `PeerID`s and `multiaddrs` of these peers to Client.
+8. Client sends a DHT lookup request for `CID` to the closest peers in XOR distance to `Hash(MH)` that it received.
+9. Go to 4.
+10. The DHT servers storing the Provider Record(s) associated with `MH` send them to Client. (Currently, if a Provider Record has been published less than 30 min before being requested, the DHT servers also send the `multiaddresses` of the Content Provider to Client).
+11. If the response from the DHT server doesn't include the `multiaddrs` associated with the Content Providers' `PeerID`s, Client performs a DHT `FindPeer` request to find the `multiaddrs` of the returned `PeerID`s.
+12. Client sends a Bitswap request for `CID` to the Content Provider (known `PeerID` and `multiaddrs`).
+13. Content Provider sends the requested content back to Client.
+
+### Double Hash DHT design
+
+**Publish Process**
+1. Content Provider wants to publish some content with identifier `CID`.
+2. Content Provider computes `HASH2`$\leftarrow{}$`SHA256(bytes("CR_DOUBLEHASH") || MH)` (`MH` is the MultiHash included in the CID).
+3. Content Provider starts a DHT lookup request for the 20 closest `PeerID`s in XOR distance to `HASH2`.
+4. Content Provider encrypts its own `PeerID` (`CPPeerID`) with `MH`, using AES-GCM. `EncPeerID = varint || Nonce || AESGCM(MH, CPPeerID || Nonce)`
```

Comment:
That is the `varint` associated with the encryption algorithm, AESGCM. This specific `varint` doesn't exist yet, we have to make a proposal on the [multicodec](https://github.com/multiformats/multicodec) repo.

Reply:
We do have a multicodec encoded as a part of `HASH2` - `DBL_SHA_256`. We also have `CR_DOUBLEHASH` appended to the multihash before it gets hashed second time. Maybe we can use either of those as a "definition of the scheme"? If we need to rotate - we can use a different prefix that would produce different `HASH2` values that would point to different encrypted records with a new encryption function. That would allow us to avoid hassle with adding a new multicodec and would also save us a few bytes per record that is substantial over a large number of CIDs.  Wdyt?

Reply:
Having a different Kademlia identifier (`HASH2`) for different encryption algorithm means that when a client wants to find the Provider Record associated with some content, the client must know which encryption algorithm (associated with a constant prefix) has been used by the Content Provider. During a migration period, from one algorithm to another, the number of DHT requests is expected to double (all client look for old and new `HASH2`). This solution reduces the `Publish` and `LookupResponse` messages by 2-3 bytes, as `EncPeerID` doesn't contain the 2-3 bytes `varint` anymore.

On the other hand, using a `varint` allows Clients to perform a single DHT lookup during encryption algorithm migrations, and allows to have multiple Provider Record encryption algorithms in use concurrently.

So the tradeoff here is:
- having 2-3 bytes less on `Publish` message (send for each republish, at the moment every 24h), and 2-3 bytes less for each Provider Record matching the requested `Prefix` (for `k = 8`, we expect ~24 bytes less per DHT Server response containing Provider Records)
- VS.
- having all encrypted Provider Records associated with the same content published at the same keyspace location, an information about the used encryption algorithm, and the possibility to have multiple Provider Record encryption algorithms concurrently all accessible in a single DHT request

Another option could be to add a 1-byte `varint` local to Double Hashing to reduce the size of the payload while having all Provider Records associated with the same content in one place.

Also, if we were to get rid of the `varint` it would IMO make sense to choose a constant string reflecting the encryption algorithm (e.g `CR_DOUBLEHASH_AESGCM`).

Reply:
> having 2-3 bytes less on Publish message (send for each republish, at the moment every 24h), and 2-3 bytes less for each Provider Record matching the requested Prefix (for k = 8, we expect ~24 bytes less per DHT Server response containing Provider Records)

We need to keep in mind how much extra storage space is going to be required. Right now cid.contact has about 1 trillion CIDs indexed and ingests 9 billion new CIDs per week. Each CID available across one or more providers - lets say 3 on average. If we were to add a new prefix right now that would require an extra 9TiBs (3 bytes * 3 providers * 1_000_000_000_000 cids).

> having all encrypted Provider Records associated with the same content published at the same keyspace location, an information about the used encryption algorithm, and the possibility to have multiple Provider Record encryption algorithms concurrently all accessible in a single DHT request

If we just replace an old value with a new one then that would require a coordinated client upgrade as old clients wouldn't know about new encryption function. So most likely we will have to return both values and leave it up to the client to pick the right one.  Also, what would be a likelihood of needing to rotate an encryption function without having to rotate hashing too? 

> Also, if we were to get rid of the varint it would IMO make sense to choose a constant string reflecting the encryption algorithm (e.g CR_DOUBLEHASH_AESGCM)

Do you mean something like that: `enc(MH, CR_DOUBLEHASH_AESGCM || PeerID)`?

Reply:
I agree that when 9TiBs seems large, but I think that it makes more sense to compare with overhead percentage rather than absolute numbers.

AESGCM nonce: 12 bytes ([source](https://crypto.stackexchange.com/questions/41601/aes-gcm-recommended-iv-size-why-12-bytes))
`MH` (used both in the `CID` and `PeerID`): 32 bytes
`TS`: 4 bytes
`Signature` (ed25519): 64 bytes
`varint`: 3 bytes

Minimal size of 1 entry in the Indexers IIUC not including Signature, nor context ID (`AESGCM nonce` + `MH_CID` + `MH_PeerID`): 12 + 32 + 32 = 76 bytes
Additional `varint` overhead: 3/76 = 3.95%

Full size (`AESGCM nonce` + `MH_CID` + `MH_PeerID` + `TS` + `Signature`): 12 + 32 + 32 + 4 + 64 = 144 bytes
Additional `varint` overhead: 3/144 = 2.08%

Reply:
> If we just replace an old value with a new one then that would require a coordinated client upgrade as old clients wouldn't know about new encryption function. So most likely we will have to return both values and leave it up to the client to pick the right one.

Exactly! During the migration period the client has to look for both the old and the new Provider Records. However if they are stored at the same location in the DHT, only 1 request is required to get both the old and new encrypted Provider Records. If the old and the new are stored in different locations, 2 requests are required to get both Provider Records.

>  Also, what would be a likelihood of needing to rotate an encryption function without having to rotate hashing too?

I think that it is unrelated. If the selected encryption algorithm becomes vulnerable to a serious attack, we need to replace it and we can keep the hash function. Hopefully, AESGCM will not be proven vulnerable anytime soon, and unplanned migrations won't be necessary.

Generally we don't want change to happen, but if it must happen we'd better be ready. Eventually all system will evolve (or die).

Reply:
> Do you mean something like that: `enc(MH, CR_DOUBLEHASH_AESGCM || PeerID)`?

Yes, or more precisely `enc(MH, "CR_DOUBLEHASH_AESGCM" || PeerID)`

Reply:
Ok thanks for your answers - that sounds reasonable. If we have `varint`  then `CR_DOUBLEHASH_AESGCM` prefix won't be necessary.